### PR TITLE
Added .xreplace() to Vector and Dyadic

### LINKED ISSUE #20445: Add .xreplace() method to sympy.physics.vector.Vector and Dyadic
Vector and Dyadic class is missing  many of the core SymPy methods. https://github.com/sympy/sympy/issues/18064#issue-539183972
xreplace() is next to be added. https://github.com/sympy/sympy/pull/20396#issuecomment-729508136

### PR DESCRIPTION
**Pull request by sidhu1012**
#### References to other Issues or PRs

Fixes #20445 

#### Brief description of what is fixed or changed
Physics Vector and Dyadic is missing core sympy methods. This PR adds .xreplace() to these classes.

#### Release Notes

* physics.vector
    * Added .xreplace() to Vector Class.
    * Added .xreplace() to Dyadic Class.

### CODE REVIEW

**Comment by sidhu1012** [comment]
@moorepants It's ready for review.

**Comment by sidhu1012** [comment]
@moorepants Shall I start working to add .simplify method, while this PR is in review?

File: sympy/physics/vector/dyadic.py [hunk 1]
```diff
@@ -539,6 +539,50 @@ def _eval_evalf(self, prec):
...
+        ========
+
+        >>> from sympy import symbols, pi
+        >>> from sympy.physics.vector import ReferenceFrame
+        >>> N = ReferenceFrame('N')
+        >>> x, y, z = symbols('x y z')
+        >>> ((1 + x*y) * (N.x | N.x)).xreplace({x: pi})
+        (pi*y + 1)*(N.x|N.x)
+        >>> ((1 + x*y) * (N.x | N.x)).xreplace({x: pi, y: 2})
+        (1 + 2*pi)*(N.x|N.x)
```
Comment (moorepants) [inline sympy/physics/vector/dyadic.py:567]: Avoid the use of `|` for operations in the documentation. I plan to deprecate these overloaded operations eventually.

File: sympy/physics/vector/tests/test_dyadic.py [hunk 2]
```diff
@@ -104,3 +104,15 @@ def test_dyadic_evalf():
...
+
+def test_dyadic_xreplace():
+    x, y, z = symbols('x y z')
+    N = ReferenceFrame('N')
+    v = x*y * (N.x | N.x)
+    assert v.xreplace({x : cos(x)}) == cos(x)*y * (N.x | N.x)
+    assert v.xreplace({x*y : pi}) == pi * (N.x | N.x)
+    v = (x*y)**z * (N.x | N.x)
+    assert v.xreplace({(x*y)**z : 1}) == (N.x | N.x)
+    assert v.xreplace({x:1, z:0}) == (N.x | N.x)
```
Comment (moorepants) [inline sympy/physics/vector/tests/test_dyadic.py:116]: Don't add new code that makes use of the overloaded vector and dyadic operations `|` or `^`. These will be deprecated. Use the outer product method in this case.

File: sympy/physics/vector/vector.py [hunk 3]
```diff
@@ -719,6 +719,48 @@ def _eval_evalf(self, prec):
             new_args.append([mat.evalf(n=prec_to_dps(prec)), frame])
         return Vector(new_args)
 
+    def xreplace(self, rule):
+        """
+        Replace occurrences of objects within the expression.
```
Comment (moorepants) [inline sympy/physics/vector/vector.py:724]: "within the measure numbers of the vector"

File: sympy/physics/vector/vector.py [hunk 4]
```diff
@@ -719,6 +719,48 @@ def _eval_evalf(self, prec):
...
+        Parameters
+        ==========
+
+        rule : dict-like
+            Expresses a replacement rule
+
+        Returns
+        =======
+
+        xreplace : the result of the replacement
```
Comment (moorepants) [inline sympy/physics/vector/vector.py:735]:
```rst
Returns
=======
Vector
  Result of the replacement.
```

File: sympy/physics/vector/dyadic.py [hunk 5]
```diff
@@ -539,6 +539,50 @@ def _eval_evalf(self, prec):
...
+        Parameters
+        ==========
+
+        rule : dict-like
+            Expresses a replacement rule
+
+        Returns
+        =======
+
+        xreplace : the result of the replacement
```
Comment (moorepants) [inline sympy/physics/vector/dyadic.py:555]:
```rst
Returns
=======
Dyadic
   Result of the replacement.
```

**Comment by moorepants** [comment]
> @moorepants Shall I start working to add .simplify method, while this PR is in review?

I believe they already have `.simplify()`.

**Comment by sidhu1012** [comment]
> I believe they already have `.simplify()`.

What's next to be added? I'm up for it!

**Comment by moorepants** [comment]
> What's next to be added? I'm up for it!

If you want to work on other things, I recommend finding open issues in the issue tracker.

**Comment by sidhu1012** [comment]
> If you want to work on other things, I recommend finding open issues in the issue tracker.

Yes, I'm doing that. Here https://github.com/sympy/sympy/issues/18064#issue-539183972 you said more core methods are to be added, so I asked.

**Comment by moorepants** [comment]
> Yes, I'm doing that. Here #18064 (comment) you said more core methods are to be added, so I asked.

I don't know others off the top of my head. You can always add everything that an expression has to make them 1-to-1 with expressions, but I'm not sure its worth it until someone presents a use case.

File: sympy/physics/vector/dyadic.py [hunk 6]
```diff
@@ -539,6 +539,51 @@ def _eval_evalf(self, prec):
...
+        Parameters
+        ==========
+
+        rule : dict-like
+            Expresses a replacement rule.
+
+        Returns
+        =======
+
+        Dyadic : Result of the replacement.
```
Comment (moorepants) [inline sympy/physics/vector/dyadic.py:555]: This still doesn't quite follow numpydoc conventions: https://numpydoc.readthedocs.io/en/latest/format.html#sections

**Comment by sidhu1012** [comment]
@moorepants Please review, I have done the reviewed changes.

**Comment by sidhu1012** [comment]
Closing and opening to restart tests.

**Comment by sidhu1012** [comment]
@moorepants Please Review.

**Comment by moorepants** [comment]
Thanks @sidhu1012! LGTM.

**Pull request merged**